src/EditorFeatures/Core/Diagnostics/AbstractPushOrPullDiagnosticsTaggerProvider.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Editor.Shared.Tagging;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Workspaces;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.CodeAnalysis.Diagnostics;
+
+internal static class DiagnosticTaggingOptions
+{
+    public static readonly Option2<bool> PullDiagnosticTagging = new(
+        "DiagnosticTaggingOptions", "PullDiagnosticTagging", defaultValue: true,
+        new FeatureFlagStorageLocation("Roslyn.PullDiagnosticTagging"));
+}
+
+/// <summary>
+/// Base type of all diagnostic taggers (classification, squiggles, suggestions, inline-diags).  Subclasses can control
+/// things by overriding functionality in this type.  Internally, this will switch to either a pull or push cased
+/// approach at instantiation time depending on our internal feature flag.
+/// </summary>
+internal abstract partial class AbstractPushOrPullDiagnosticsTaggerProvider<TTag> : ITaggerProvider
+    where TTag : ITag
+{
+    private readonly ITaggerProvider _underlyingTaggerProvider;
+
+    protected readonly IGlobalOptionService GlobalOptions;
+
+    protected AbstractPushOrPullDiagnosticsTaggerProvider(
+        IThreadingContext threadingContext,
+        IDiagnosticService diagnosticService,
+        IDiagnosticAnalyzerService analyzerService,
+        IGlobalOptionService globalOptions,
+        ITextBufferVisibilityTracker? visibilityTracker,
+        IAsynchronousOperationListener listener)
+    {
+        GlobalOptions = globalOptions;
+
+        if (globalOptions.GetOption(DiagnosticTaggingOptions.PullDiagnosticTagging))
+        {
+            _underlyingTaggerProvider = new PullDiagnosticsTaggerProvider(
```

Comment:
can you also confirm LSP isn't running when the pull tagger is on?  Not seeing anything obvious disabling the tagger when LSP is active

Reply:
so this is tricksy (and i can comment).  LSP pull does *not* disable Tagger-Pull (or push) *overall*.  Consider the case of 'inline diagnostics'. They are not replaced with LSP diagnostics.  So they must continue to run.  So, instead of checking that flag in one place, we check in each feature.  Classification/Squiggles/Suggestion-tags all defer to LSP if LSP is on.  But inline-diagnostics does not care.